multi: replace `LockOutpoint` with `LeaseOutput`

aakselrod opened the pull request:
## Change Description
This fixes #8476.

## Steps to Test
The first commit, containing an integration test, is sufficient to test the change. The integration test fails when run prior to merging the subsequent commits which fix the issue, and passes after.

## Pull Request Checklist
### Testing
- [x] Your PR passes all CI checks.
- [x] Tests covering the positive and negative (error paths) are included.
- [x] Bug fixes contain tests triggering the bug to prevent regressions.

### Code Style and Documentation
- [x] The change obeys the [Code Documentation and Commenting](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#CodeDocumentation) guidelines, and lines wrap at 80.
- [x] Commits follow the [Ideal Git Commit Structure](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#IdealGitCommitStructure).
- [x] Any new logging statements use an appropriate subsystem and logging level.
- [x] Any new lncli commands have appropriate tags in the comments for the rpc in the proto file.
- [x]  [There is a change description in the release notes](https://github.com/lightningnetwork/lnd/tree/master/docs/release-notes), or `[skip ci]` in the commit message for small changes.

📝 Please see our  [Contribution Guidelines](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md) for further guidance.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- **New Features**
	- Introduced an enhanced output leasing mechanism for more efficient handling of channel funding processes.

- **Refactor**
	- Updated and unified the output locking mechanism across the system to use the new `OutputLeaser` interface.
	- Removed obsolete locking and unlocking functionalities from various components.

- **Tests**
	- Added new test cases to verify the behavior of locked balance during channel opening processes.

- **Chores**
	- Cleaned up unused import statements and outdated references to improve code maintainability and clarity.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

aakselrod commented on itest/lnd_open_channel_test.go:
```diff
@@ -822,3 +822,53 @@ func testSimpleTaprootChannelActivation(ht *lntest.HarnessTest) {
+	require.NoError(ht, err)
+
+	balance = alice.RPC.WalletBalance()
+	require.NotEqual(ht, int64(0), balance.LockedBalance)
```
This is where the itest fails before the fix. `LockedBalance` only counts outputs that are locked via `LeaseOutput`, leaving out ones that are locked via `LockOutpoint`. This results in 0 locked balance, though funds are reserved for opening the channel.

aakselrod commented:
@coderabbitai review

aakselrod commented on sweep/walletsweep.go:
```diff
@@ -134,17 +137,18 @@ type CoinSelectionLocker interface {
+// OutputLeaser allows a caller to lease/release an output. When leased, the
+// outputs shouldn't be used for any sort of channel funding or coin selection.
+// Leased outputs are expected to be persisted between restarts.
+type OutputLeaser interface {
+	// LeaseOutput leases a target output, rendering it unusable for coin
 	// selection.
-	LockOutpoint(o wire.OutPoint)
+	LeaseOutput(i wtxmgr.LockID, o wire.OutPoint, d time.Duration) (
+		time.Time, []byte, btcutil.Amount, error)
 
-	// UnlockOutpoint unlocks a target outpoint, allowing it to be used for
+	// ReleaseOutput releases a target output, allowing it to be used for
 	// coin selection once again.
-	UnlockOutpoint(o wire.OutPoint)
+	ReleaseOutput(i wtxmgr.LockID, o wire.OutPoint) error
```
The interface is just a small subset of the `lnwallet.WalletController` interface. Maybe a (future?) refactor could move this interface to `chanfunding` or so, and then use the same interface from everywhere it's needed, but I'm not sure just adding more comments here specifically would be any more useful.

aakselrod commented on sweep/walletsweep.go:
```diff
@@ -179,11 +183,11 @@
 // leftover amount after these outputs and transaction fee, is sent to a single
 // output, as specified by the change address. The sweep transaction will be
 // crafted with the target fee rate, and will use the utxoSource and
-// outpointLocker as sources for wallet funds.
+// outputLeaser as sources for wallet funds.
 func CraftSweepAllTx(feeRate, maxFeeRate chainfee.SatPerKWeight,
 	blockHeight uint32, deliveryAddrs []DeliveryAddr,
 	changeAddr btcutil.Address, coinSelectLocker CoinSelectionLocker,
-	utxoSource UtxoSource, outpointLocker OutpointLocker,
+	utxoSource UtxoSource, outputLeaser OutputLeaser,
 	signer input.Signer, minConfs int32) (*WalletSweepPackage, error) {
 
 	// TODO(roasbeef): turn off ATPL as well when available?
```
Do we want to log errors here? Might be a good idea...

aakselrod replied:
Added error logging.

aakselrod commented on sweep/walletsweep.go:
```diff
@@ -219,7 +227,13 @@
 		// crafting out sweep all transaction.
 		for _, utxo := range utxos {
-			outpointLocker.LockOutpoint(utxo.OutPoint)
+			_, _, _, err = outputLeaser.LeaseOutput(
+				chanfunding.LndInternalLockID, utxo.OutPoint,
+				chanfunding.DefaultLockDuration,
+			)
+			if err != nil {
+				return err
+			}
```
I believe this error would be logged by an upstream caller when a lease fails.

aakselrod reviewed (commented):
Responding to coderabbit comments.

On a higher level, we don't need `volatile_locked_balance` or `ListLockedOutpoints` as mentioned in #8476 since this PR eliminates volatile locks altogether.

guggero commented on itest/lnd_open_channel_test.go:
```diff
@@ -822,3 +822,53 @@ func testSimpleTaprootChannelActivation(ht *lntest.HarnessTest) {
+	// the funding flow can be properly executed.
+	alice := ht.Alice
+	bob := ht.Bob
+	ht.EnsureConnected(alice, bob)
```
nit: this should have already been asserted by the framework itself.

aakselrod replied:
Removed this line.

guggero commented on itest/lnd_open_channel_test.go:
```diff
@@ -822,3 +822,53 @@ func testSimpleTaprootChannelActivation(ht *lntest.HarnessTest) {
+
+	// We first make sure Alice has no locked wallet balance.
+	balance := alice.RPC.WalletBalance()
+	require.Equal(ht, int64(0), balance.LockedBalance)
```
nit: could use `require.EqualValues()` instead of explicit cast.

aakselrod replied:
Used `EqualValues` and `NotEqualValues` instead of casts, and checked that the itest fails correctly before the changes, and passes after the changes.

guggero commented on lnwallet/chanfunding/wallet_assembler.go:
```diff
@@ -488,7 +488,13 @@ func (w *WalletAssembler) ProvisionChannel(r *Request) (Intent, error) {
-			w.cfg.CoinLocker.LockOutpoint(outpoint)
+			_, _, _, err = w.cfg.CoinLeaser.LeaseOutput(
+				LndInternalLockID, outpoint,
+				DefaultLockDuration,
```
I wonder: Since before this change the lock would basically exist forever (or until the next restart of the node), should we also use an "infinite" duration here? Or do we kind of rely on the general funding timeout we have in `lnd` that is 10 minutes?

aakselrod replied:
I think we have an interplay of several factors:

- the `chanfunding.DefaultLockDuration` (formerly `walletrpc.DefaultLockDuration`) is 10m
- the `lncfg.DefaultReservationTimeout` is 10m
- only the dev config allows changing the reservation timeout from 10m, which is used in itests
- the reservation timeout applies to time since last update to the reservation, rather than time since the reservation was created - this means it can go longer than 10m if the peer is slow enough to respond
- the zombie sweeper iterates through `activeReservations` which I believe is NOT populated on startup from any non-volatile data, so it would miss releasing outputs with infinite expiration and would cause them to need to be released manually in the case that LND restarts with an active reservation

This makes me think that maybe the duration should be the reservation timeout multiplied by a constant factor for the number of state transitions we expect, to allow the peer to idle up to the timeout at each step. Since the timeout is configurable in the dev environment, perhaps we pass that value to lnwallet when initializing it as well?

guggero replied:
Ah, right, we have the zombie sweeper that does the cleanup. Maybe we could add a `Cleanup()` method to the reservation that would be called by the zombie sweeper to release the locks immediately? That combined with a multiplication factor of the reservation timeout sounds good to me.

aakselrod replied:
- The zombie sweeper calls `failFundingFlow()` on the manager with the channel ID here: https://github.com/lightningnetwork/lnd/blob/72764b14734870d03a1212795e2ec7c02f3b4500/funding/manager.go#L4848
- which calls `cancelReservationCtx()` on the manager with the channel ID here: https://github.com/lightningnetwork/lnd/blob/72764b14734870d03a1212795e2ec7c02f3b4500/funding/manager.go#L912
- which calls `Cancel()` on the `ChannelReservation` here: https://github.com/lightningnetwork/lnd/blob/72764b14734870d03a1212795e2ec7c02f3b4500/funding/manager.go#L4885
- which sends a `fundingReserveCancelMsg` to the wallet's `msgChan` with the reservation ID here: https://github.com/lightningnetwork/lnd/blob/72764b14734870d03a1212795e2ec7c02f3b4500/lnwallet/reservation.go#L795
- which calls `handleFundingCancelRequest()` on the wallet with the message here: https://github.com/lightningnetwork/lnd/blob/72764b14734870d03a1212795e2ec7c02f3b4500/lnwallet/wallet.go#L620
- which calls `UnlockOutpoint()` here: https://github.com/lightningnetwork/lnd/blob/72764b14734870d03a1212795e2ec7c02f3b4500/lnwallet/wallet.go#L1428
- which is replaced in this PR already with a `ReleaseOutput()`

So I think the `Cleanup()` part of this might be redundant, as the zombie sweeper already calls cleanup code that this PR modifies to work with leases instead of locks. I think all I need to do is update this timeout.

guggero replied:
Okay cool. Sounds good to me, thanks for confirming!

aakselrod replied:
Thanks! I got most of the way there today, will finish it up tomorrow.

aakselrod replied:
Looking more closely at the multiplier to use:

- Channel establishment v1 should have a multiplier of 1 (the time between `funding_created` and `funding_signed` messages)
- Channel establishment v2 specifically recommends not locking UTXOs and reusing them between channel open attempts for dual-funded channels, to help [prevent liquidity griefing](https://github.com/lightning/bolts/blob/master/02-peer-protocol.md#liquidity-griefing). In addition, since there's a loop that's iterated for each input added (from each side), the multiplier isn't constant.
 
For itests, the reservation timeout is shorter, so using the longer value of the default timeout is safe. I think it's safe to just use `lncfg.DefaultReservationTimeout` in this case.

aakselrod replied:
Moved `lncfg.DefaultReservationTimeout` to `lnwallet/chanfunding` to avoid import cycles, updated all references to it, and used it explicitly instead of `DefaultLockDuration` for channel reservations.

aakselrod replied:
Ah, forgot to actually use `DefaultReservationTimeout` instead of `DefaultLockDuration`. Fixed.

guggero reviewed (commented):
Very nice cleanup/consolidation of behavior! Code looks good to me, main question is around what timeout to use.

aakselrod commented on itest/lnd_open_channel_test.go:
```diff
@@ -822,3 +822,53 @@ func testSimpleTaprootChannelActivation(ht *lntest.HarnessTest) {
+	// Finally, we check to make sure the balance is unlocked again.
+	balance = alice.RPC.WalletBalance()
+	require.Equal(ht, int64(0), balance.LockedBalance)
+}
```
Completed TODO.

aakselrod commented:
Pushed an update:
- Updated release notes
- Fixed nits and comments in itest
- Added error logging (as warnings) to output release loops in `lnwallet/chanfunding` and `sweep` packages
- Moved `DefaultReservationTimeout` to `lnwallet/chanfunding` and used it explicitly in wallet assembler

guggero commented on lnwallet/chanfunding/wallet_assembler.go:
```diff
@@ -219,7 +219,13 @@ func (f *FullIntent) Outputs() []*wire.TxOut {
+			LndInternalLockID, coin.OutPoint,
+		)
+		if err != nil {
+			log.Warnf("Failed to release UTXO %s (%s))",
```
nit: use `%v` for errors?

aakselrod replied:
Done! :heavy_check_mark:

guggero reviewed (approved):
Very nice, LGTM :tada:

lightninglabs-deploy commented:
@yyforyongyu: review reminder
@ziggie1984: review reminder
@aakselrod, remember to re-request review from reviewers when ready

aakselrod commented:
> @yyforyongyu: review reminder @ziggie1984: review reminder @aakselrod, remember to re-request review from reviewers when ready

Just need one more approval! :pray:

yyforyongyu reviewed (commented):
Looking good! Just a few comments on the itest then this should be good to go!

aakselrod commented on itest/lnd_open_channel_test.go:
```diff
@@ -822,3 +822,52 @@ func testSimpleTaprootChannelActivation(ht *lntest.HarnessTest) {
+// made for opening a channel, the balance of the required outputs shows
+// up as locked balance in the WalletBalance response.
+func testOpenChannelLockedBalance(ht *lntest.HarnessTest) {
+	// Before we start the test, we'll ensure both sides are connected so
```
Thanks for the catch! I removed this comment, and rearranged the vars a little bit to support the `wait.NoError` change below.

aakselrod commented on itest/lnd_open_channel_test.go:
```diff
@@ -822,3 +822,52 @@ func testSimpleTaprootChannelActivation(ht *lntest.HarnessTest) {
+
+	// After that, we receive the request on Bob's side, get the wallet
+	// balance from Alice, and ensure the locked balance is non-zero.
+	req, err := stream.Recv()
```
Done on both!

yyforyongyu commented:
This should be good to do, but I'm having trouble understanding why this itest failed,
```
harness_rpc.go:97: 
        	Error Trace:	/home/runner/work/lnd/lnd/lntest/rpc/harness_rpc.go:97
        	            				/home/runner/work/lnd/lnd/lntest/rpc/lnd.go:374
        	            				/home/runner/work/lnd/lnd/itest/lnd_psbt_test.go:1538
        	            				/home/runner/work/lnd/lnd/itest/lnd_psbt_test.go:1019
        	            				/home/runner/work/lnd/lnd/itest/lnd_psbt_test.go:661
        	            				/home/runner/work/lnd/lnd/lntest/harness.go:353
        	            				/home/runner/work/lnd/lnd/itest/lnd_test.go:139
        	Error:      	Received unexpected error:
        	            	rpc error: code = DeadlineExceeded desc = context deadline exceeded
        	Messages:   	Alice: failed to call SendCoins
```

It seems that postgres was really slow, just wanna be sure it wasn't introduced by this PR.

aakselrod commented:
It looks like it might actually have to do with this PR. The `send_all` option used to use in-memory locks, but now uses leases, which is definitely slower. I wonder if moving the failing test will help until #8529 goes in? I'll try it in a new commit and see what happens.

aakselrod commented:
That seems to be worse. Going to force-push to remove that commit.

aakselrod commented:
I'm going to explore more locally and see if I can reproduce/debug the issue in this test.

aakselrod commented:
Those itests seem to pass locally for me. Curiouser and curiouser...

guggero commented:
Maybe it's a matter of available resources on the GitHub runner? It seems to me like they are being scaled down somewhat, at least it seems the available disk space is less than before. So maybe CPU is also being throttled?
We could try increasing the DB specific timeout here: https://github.com/lightningnetwork/lnd/blob/d5af76cb2973b92ead1f602baba2f71e8e939cc3/lntest/wait/timeouts_remote_db.go#L23

aakselrod commented:
Ah, yeah, thus the cross compile check running out of space. I'll increase the timeout and see how the tests do!

aakselrod commented:
I think the [`unit-race` failures](https://github.com/lightningnetwork/lnd/actions/runs/8300573490/job/22718808474?pr=8496) are unrelated to this PR.

aakselrod commented:
Looks like the postgres itests passed with a longer timeout.

yyforyongyu reviewed (approved):
LGTM🎉
